File: impl/java/src/main/java/com/cptnfizzbin/keycard/policy/Policy.java
package com.cptnfizzbin.keycard.policy;

import com.cptnfizzbin.keycard.KeycardConfig;
import com.cptnfizzbin.keycard.action.Action;
import com.cptnfizzbin.keycard.conditions.ConditionResolver;
import com.cptnfizzbin.keycard.conditions.OperatorCatalog;
import com.cptnfizzbin.keycard.errors.PolicyException;
import com.cptnfizzbin.keycard.errors.PolicyLoadException;
import com.cptnfizzbin.keycard.errors.PolicyVersionException;
import com.cptnfizzbin.keycard.lib.Catalog;
import com.cptnfizzbin.keycard.subject.Subject;
import com.cptnfizzbin.keycard.version.KeyCardVersion;
import lombok.Getter;
import lombok.val;
import org.semver4j.Semver;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

public final class Policy {
    @Getter
    private final PolicyDefinition definition;

    private final ConditionResolver resolver;
    private final KeycardConfig config;
    private final Map<String, String> actionReverseMap;
    private final Map<String, String> subjectReverseMap;

    // Snapshotted at construction, so evaluation only ever sees the rules and
    // wildcard tokens that were validated here - later mutation of the
    // (mutable) PolicyDefinition can't bypass validation.
    private final List<PolicyDefinition.Rule> rules;
    private final WildcardToken anyAction;
    private final WildcardToken anySubject;

    public Policy(PolicyDefinition definition) {
        this(definition, new KeycardConfig());
    }

    public Policy(PolicyDefinition definition, KeycardConfig config) {
        validateVersion(definition.version());
        this.definition = definition;
        this.config = config;
        this.resolver = new ConditionResolver(config.operators());

        Catalog.Resolution actions = Catalog.build(null, config.actions(), Action::name, "action");
        Catalog.Resolution subjects = Catalog.build(null, config.subjects(), Subject::name, "subject");
        this.actionReverseMap = actions.reverseMap();
        this.subjectReverseMap = subjects.reverseMap();

        this.rules = definition.getRules();
        this.anyAction = Wildcards.effectiveAnyAction(definition.meta());
        this.anySubject = Wildcards.effectiveAnySubject(definition.meta());

        // Structural validity is what evaluation itself relies on, so it's
        // checked regardless of emitMeta - only the catalog checks are gated.
        validateRuleShapes(rules, anyAction, anySubject);
        if (config.emitMeta()) {
            validateOperatorsRegistered(definition, resolver);
            validateRuleCatalogs(definition, rules, anyAction, anySubject, actions.names(), subjects.names());
        }
    }

    /**
     * A bare-type check (no instance): a conditional rule can never match this.
     */
    public boolean can(Action action, Subject<?, ?> subject) {
        return checkPermission(action, subject);
    }

    public boolean cannot(Action action, Subject<?, ?> subject) {
        return !can(action, subject);
    }

    public void require(Action action, Subject<?, ?> subject) throws PolicyException {
        if (!can(action, subject)) {
            String actionName = Catalog.resolveName(actionReverseMap, action.name());
            String subjectName = Catalog.resolveName(subjectReverseMap, subject.name());
            throw new PolicyException("Access denied: cannot " + actionName + " on " + subjectName);
        }
    }

    /**
     * Reverse scan over `rules`, returning the effect of
     * the first (i.e. most-recently-declared) rule whose action, subject,
     * and (if present) conditions all match. There is no independent
     * "allow AND NOT deny" veto and no combination of multiple matching
     * rules: exactly one rule decides the outcome, or none does and the
     * result is default deny.
     */
    private boolean checkPermission(Action action, Subject<?, ?> subject) {
        String actionName = Catalog.resolveName(actionReverseMap, action.name());
        String subjectName = Catalog.resolveName(subjectReverseMap, subject.name());

        for (int i = rules.size() - 1; i >= 0; i--) {
            PolicyDefinition.Rule rule = rules.get(i);

            if (!Wildcards.matches(actionName, rule.action(), anyAction)) continue;
            if (!Wildcards.matches(subjectName, rule.subjectName(), anySubject)) continue;

            Map<String, Object> conditions = rule.conditions();
            if (conditions != null) {
                // A conditional rule can never be satisfied by a bare-type/no-instance
                // check - there's no instance data for the condition to inspect.
                if (subject.claims().isEmpty()) continue;
                if (!resolver.evaluate(subject.claims().get(), conditions)) continue;
            }

            return "allow".equals(rule.effect());
        }

        return false;
    }

    private static void validateVersion(String version) {
        val supported = Optional.ofNullable(Semver.coerce(version))
            .orElseThrow(() -> new PolicyVersionException("Invalid version " + version))
            .satisfies(KeyCardVersion.KEYCARD_POLICY_SUPPORTED_VERSIONS);

        if (!supported) {
            throw new PolicyVersionException(
                "Unsupported policy version \"" + version + "\": this implementation supports " + KeyCardVersion.KEYCARD_POLICY_SUPPORTED_VERSIONS
            );
        }
    }

    /**
     * when `meta.operators` is declared, every
     * name it lists MUST already be registered on this Policy - built-in or
     * custom - checked once here when loading a policy, regardless of
     * whether any rule actually reaches that operator during evaluation.
     */
    private static void validateOperatorsRegistered(PolicyDefinition definition, ConditionResolver resolver) {
        PolicyDefinition.Meta meta = definition.meta();
        List<String> declared = meta != null ? meta.operators() : null;
        if (declared == null) return;

        resolver.assertAllRegistered(declared);
    }

    /**
     * Malformed rule tuples and conditional both-sides-wildcarded rules -
     * always checked, since evaluation can't proceed safely past either.
     */
    private static void validateRuleShapes(List<PolicyDefinition.Rule> rules, WildcardToken anyAction, WildcardToken anySubject) {
        for (PolicyDefinition.Rule rule : rules) {
            if (!"allow".equals(rule.effect()) && !"deny".equals(rule.effect())) {
                throw new PolicyLoadException(
                    "Malformed rule tuple: effect must be \"allow\" or \"deny\", got " + rule.effect() + "."
                );
            }
            if (rule.action() == null) {
                throw new PolicyLoadException("Malformed rule tuple: action must be a string, got null.");
            }
            if (rule.subjectName() == null) {
                throw new PolicyLoadException("Malformed rule tuple: subject must be a string, got null.");
            }

            if (isWildcard(anyAction, rule.action()) && isWildcard(anySubject, rule.subjectName()) && rule.conditions() != null) {
                throw new PolicyLoadException(
                    "Rule [" + rule.effect() + ", " + rule.action() + ", " + rule.subjectName()
                        + "] is wildcarded on both the action and the subject but carries a Conditions element - this MUST be unconditional."
                );
            }
        }
    }

    private static boolean isWildcard(WildcardToken any, String value) {
        return any instanceof WildcardToken.Named named && value.equals(named.token());
    }

    /**
     * @param configActionNames resolved catalog names (see {@code lib.Catalog}) that, when given, widen the `meta.actions` catalog below beyond what `definition.meta` declares.
     * @param configSubjectNames resolved catalog names (see {@code lib.Catalog}) that, when given, widen the `meta.subjects` catalog below beyond what `definition.meta` declares.
     */
    private static void validateRuleCatalogs(
        PolicyDefinition definition,
        List<PolicyDefinition.Rule> rules,
        WildcardToken anyAction,
        WildcardToken anySubject,
        List<String> configActionNames,
        List<String> configSubjectNames
    ) {
        PolicyDefinition.Meta meta = definition.meta();

        List<String> metaActions = meta != null ? meta.actions() : null;
        List<String> metaSubjects = meta != null ? meta.subjects() : null;
        List<String> metaOperators = meta != null ? meta.operators() : null;

        Set<String> actionsCatalog = null;
        if ((metaActions != null && !metaActions.isEmpty()) || !configActionNames.isEmpty()) {
            actionsCatalog = new LinkedHashSet<>();
            if (metaActions != null) actionsCatalog.addAll(metaActions);
            actionsCatalog.addAll(configActionNames);
        }

        Set<String> subjectsCatalog = null;
        if ((metaSubjects != null && !metaSubjects.isEmpty()) || !configSubjectNames.isEmpty()) {
            subjectsCatalog = new LinkedHashSet<>();
            if (metaSubjects != null) subjectsCatalog.addAll(metaSubjects);
            subjectsCatalog.addAll(configSubjectNames);
        }

        Set<String> operatorsCatalog = metaOperators != null ? new LinkedHashSet<>(metaOperators) : null;

        for (PolicyDefinition.Rule rule : rules) {
            boolean isWildcardAction = isWildcard(anyAction, rule.action());
            boolean isWildcardSubject = isWildcard(anySubject, rule.subjectName());

            if (actionsCatalog != null && !isWildcardAction && !actionsCatalog.contains(rule.action())) {
                throw new PolicyLoadException("Rule action \"" + rule.action() + "\" is not covered by meta.actions.");
            }
            if (subjectsCatalog != null && !isWildcardSubject && !subjectsCatalog.contains(rule.subjectName())) {
                throw new PolicyLoadException("Rule subject \"" + rule.subjectName() + "\" is not covered by meta.subjects.");
            }

            if (operatorsCatalog != null && rule.conditions() != null) {
                Set<String> used = new LinkedHashSet<>();
                collectCustomOperators(rule.conditions(), used);
                for (String op : used) {
                    if (!operatorsCatalog.contains(op)) {
                        throw new PolicyLoadException("Rule uses custom operator \"" + op + "\" not covered by meta.operators.");
                    }
                }
            }
        }
    }

    /**
     * Recursively collects every non-built-in, `$`-prefixed operator name
     * used anywhere in a Conditions tree - used to enforce `meta.operators`
     * coverage when loading a policy.
     */
    private static void collectCustomOperators(Object condition, Set<String> out) {
        if (!(condition instanceof Map<?, ?> map)) return;

        for (Map.Entry<?, ?> entry : map.entrySet()) {
            String key = String.valueOf(entry.getKey());
            Object value = entry.getValue();

            if (key.startsWith("$")) {
                if (key.equals("$or") || key.equals("$and")) {
                    if (value instanceof List<?> list) {
                        for (Object c : list) collectCustomOperators(c, out);
                    }
                } else if (key.equals("$not")) {
                    collectCustomOperators(value, out);
                } else if (key.equals("$field") && value instanceof List<?> list && list.size() == 2) {
                    collectCustomOperators(list.get(1), out);
                } else if (!OperatorCatalog.BUILTIN_NAMES.contains(key)) {
                    out.add(key);
                }
            } else {
                collectCustomOperators(value, out);
            }
        }
    }
}

File: impl/js/src/policy/policy.ts
import * as semver from "semver"

import type { PolicyDefinition, RuleTuple } from "./policyDefinition.ts"
import { DISABLED, effectiveAnyAction, effectiveAnySubject } from "./wildcards.ts"
import type { Action } from "../action/index.ts"
import type { AnyCondition } from "../conditions/condition.ts"
import { BUILTIN_OPERATOR_NAMES } from "../conditions/conditionResolver.ts"
import { ConditionResolver } from "../conditions/index.ts"
import type { AnyOperator } from "../conditions/operators/operator.ts"
import { normalizeOperators } from "../conditions/operators/operator.ts"
import { PolicyError, PolicyLoadException, PolicyVersionException } from "../errors/index.ts"
import type { KeycardConfig } from "../keycardConfig.ts"
import { buildCatalog, resolveName } from "../lib/catalog.ts"
import { getLogger } from "../lib/logger.ts"
import type { Subject } from "../subject/index.ts"
import type { SubjectFieldMapper } from "../subject/subjectFieldMapper.ts"
import { KEYCARD_POLICY_SUPPORTED_VERSIONS } from "../version.ts"

/**
 * Recursively collects every non-built-in, `$`-prefixed operator name used
 * anywhere in a Conditions tree - used to enforce `meta.operators`
 * coverage when loading a policy.
 */
function collectCustomOperators(condition: AnyCondition | undefined, out: Set<string>): void {
  if (condition === undefined || condition === null || typeof condition !== "object") return

  for (const [key, value] of Object.entries(condition)) {
    if (key.startsWith("$")) {
      if (key === "$or" || key === "$and") {
        if (Array.isArray(value)) value.forEach((c: AnyCondition) => collectCustomOperators(c, out))
      } else if (key === "$not") {
        collectCustomOperators(value, out)
      } else if (key === "$field" && Array.isArray(value) && value.length === 2) {
        collectCustomOperators(value[1], out)
      } else if (!BUILTIN_OPERATOR_NAMES.has(key)) {
        out.add(key)
      }
    } else {
      collectCustomOperators(value, out)
    }
  }
}

export class Policy<
  TActions extends Action = Action,
  TSubjects extends Subject = Subject,
  TOperators extends AnyOperator = never,
> {
  private readonly definition: PolicyDefinition
  private readonly resolver: ConditionResolver
  private readonly config: KeycardConfig<TOperators>
  private readonly actionCatalog: Map<string, string>
  private readonly subjectCatalog: Map<string, string>
  private readonly warnedDynamicIds = new Set<string>()

  /**
   * @param config shared, optional config also accepted by `PolicyBuilder`
   *   (SPEC_V0.md and the SubjectFieldMapper feature): `actions`/`subjects`
   *   widen the `meta.actions`/`meta.subjects` catalogs beyond what
   *   `definition.meta` itself declares, and double as a catalog resolving
   *   a dynamic (no-name) Action/Subject's random name to its key, built
   *   once here and cached (see `lib/catalog.ts`); `operators` - an
   *   `AnyOperator[]` or an `OperatorCatalog` - is normalized once here;
   *   `mapper` is consulted for a subject's fields whenever the `Subject`
   *   passed to {@link can} doesn't carry its own `fieldMapper`; `emitMeta`
   *   (default `true`) gates the eager catalog/operator validation below -
   *   see `KeycardConfig`'s doc.
   */
  constructor(
    definition: PolicyDefinition,
    config: KeycardConfig<TOperators> = {},
  ) {
    Policy.validateVersion(definition.version)

    const emitMeta = config.emitMeta ?? true
    const actions = buildCatalog(config.actions, "action", emitMeta)
    const subjects = buildCatalog(config.subjects, "subject", emitMeta)

    this.definition = definition
    this.config = config
    this.actionCatalog = actions.reverseMap
    this.subjectCatalog = subjects.reverseMap
    this.resolver = new ConditionResolver(normalizeOperators(config.operators))
    if (emitMeta) Policy.validateOperatorsRegistered(definition, this.resolver)
    Policy.validateRules(definition, actions.names, subjects.names, emitMeta)
  }

  /**
   * Builds a Policy from an already-parsed PolicyDefinition. KeyCard itself
   * never reads or writes policy.yaml text - an application (or a test, via
   * a YAML library of its own choosing) parses the file into a plain
   * PolicyDefinition object and hands it to KeyCard.
   */
  static from<
    TActions extends Action = Action,
    TSubjects extends Subject = Subject,
    TOperators extends AnyOperator = never,
  >(
    definition: PolicyDefinition,
    config: KeycardConfig<TOperators> = {},
  ): Policy<TActions, TSubjects, TOperators> {
    return new Policy(definition, config)
  }

  private static validateVersion(version: string): void {
    // PATCH (and MINOR) may be omitted - "1"/"1.0" are valid
    // shorthand for "1.0.0" - so coerce before comparing rather than
    // requiring a strict three-component string.
    const coerced = semver.coerce(version)
    if (!coerced || !semver.satisfies(coerced, KEYCARD_POLICY_SUPPORTED_VERSIONS)) {
      throw new PolicyVersionException(
        `Unsupported policy version "${coerced}": this implementation supports ${KEYCARD_POLICY_SUPPORTED_VERSIONS}.`,
      )
    }
  }

  /**
   * when `meta.operators` is declared, every
   * name it lists MUST already be registered on this Policy - built-in or
   * custom - checked once here when loading a policy, regardless of
   * whether any rule actually reaches that operator during evaluation.
   * Only run when `emitMeta` is true - see `KeycardConfig.emitMeta`'s doc.
   */
  private static validateOperatorsRegistered(definition: PolicyDefinition, resolver: ConditionResolver): void {
    const declared = definition.meta?.operators
    if (!declared) return

    resolver.assertAllRegistered(declared)
  }

  /**
   * @param configActionNames/@param configSubjectNames resolved catalog names (see `lib/catalog.ts`) that, when given, widen the `meta.actions`/`meta.subjects` catalogs below beyond what `definition.meta` declares.
   * @param emitMeta when false, skips only the catalog-coverage checks
   *   below (a rule action/subject/operator not covered by
   *   `meta.actions`/`meta.subjects`/`meta.operators`) - the structural
   *   checks (malformed rule tuples, EC-6) always run regardless, since
   *   those guard evaluation correctness rather than diagnostics.
   */
  private static validateRules(
    definition: PolicyDefinition,
    configActionNames: string[],
    configSubjectNames: string[],
    emitMeta: boolean,
  ): void {
    const meta = definition.meta
    const anyAction = effectiveAnyAction(meta)
    const anySubject = effectiveAnySubject(meta)

    const actionsCatalog = emitMeta && (meta?.actions || configActionNames.length > 0)
      ? new Set([...(meta?.actions ?? []), ...configActionNames])
      : undefined
    const subjectsCatalog = emitMeta && (meta?.subjects || configSubjectNames.length > 0)
      ? new Set([...(meta?.subjects ?? []), ...configSubjectNames])
      : undefined
    const operatorsCatalog = emitMeta && meta?.operators ? new Set(meta.operators) : undefined

    for (const rule of definition.rules as RuleTuple[]) {
      if (!Array.isArray(rule) || rule.length < 3) {
        throw new PolicyLoadException(
          `Malformed rule tuple (fewer than 3 elements): ${JSON.stringify(rule)}.`,
        )
      }

      const [effect, action, subjectName, conditions] = rule

      if (effect !== "allow" && effect !== "deny") {
        throw new PolicyLoadException(
          `Malformed rule tuple: effect must be "allow" or "deny", got ${JSON.stringify(effect)}.`,
        )
      }
      if (typeof action !== "string") {
        throw new PolicyLoadException(
          `Malformed rule tuple: action must be a string, got ${JSON.stringify(action)}.`,
        )
      }
      if (typeof subjectName !== "string") {
        throw new PolicyLoadException(
          `Malformed rule tuple: subject must be a string, got ${JSON.stringify(subjectName)}.`,
        )
      }

      const isWildcardAction = anyAction !== DISABLED && action === anyAction
      const isWildcardSubject = anySubject !== DISABLED && subjectName === anySubject

      if (isWildcardAction && isWildcardSubject && conditions) {
        throw new PolicyLoadException(
          `Rule [${effect}, ${action}, ${subjectName}] is wildcarded on both the action and the subject but carries a Conditions element - this MUST be unconditional (SPEC_V0.md property 5, EC-6).`,
        )
      }

      if (actionsCatalog && !isWildcardAction && !actionsCatalog.has(action)) {
        throw new PolicyLoadException(
          `Rule action "${action}" is not covered by meta.actions.`,
        )
      }
      if (subjectsCatalog && !isWildcardSubject && !subjectsCatalog.has(subjectName)) {
        throw new PolicyLoadException(
          `Rule subject "${subjectName}" is not covered by meta.subjects.`,
        )
      }

      if (operatorsCatalog && conditions) {
        const used = new Set<string>()
        collectCustomOperators(conditions, used)
        for (const op of used) {
          if (!operatorsCatalog.has(op)) {
            throw new PolicyLoadException(
              `Rule uses custom operator "${op}" not covered by meta.operators.`,
            )
          }
        }
      }
    }
  }

  def(): PolicyDefinition {
    return this.definition
  }

  can(action: TActions, subject: TSubjects): boolean {
    return this.checkPermission(action, subject)
  }

  cannot(action: TActions, subject: TSubjects): boolean {
    return !this.can(action, subject)
  }

  require(action: TActions, subject: TSubjects): void {
    if (!this.can(action, subject)) {
      const actionName = resolveName(this.actionCatalog, action.name)
      const subjectName = resolveName(this.subjectCatalog, subject.name)
      throw new PolicyError(`"${actionName}" is not allowed on this "${subjectName}"`)
    }
  }

  /**
   * SPEC_V0.md: reverse scan over `rules`, returning the effect of
   * the first (i.e. most-recently-declared) rule whose action, subject,
   * and (if present) conditions all match. There is no independent
   * "allow AND NOT deny" veto and no combination of multiple matching
   * rules: exactly one rule decides the outcome, or none does and the
   * result is default deny.
   */
  private checkPermission(action: TActions, subject: TSubjects): boolean {
    const meta = this.definition.meta
    const anyAction = effectiveAnyAction(meta)
    const anySubject = effectiveAnySubject(meta)
    const rules = this.definition.rules

    for (let i = rules.length - 1; i >= 0; i--) {
      const [effect, ruleAction, ruleSubject, ruleConditions] = rules[i]

      if (!this.matchesAction(action, ruleAction, anyAction)) continue
      if (!this.matchesSubject(subject, ruleSubject, anySubject)) continue

      if (ruleConditions) {
        // A conditional rule can never be satisfied by a bare-type/no-instance
        // check - there's no instance data for the condition to inspect (EC-7).
        if (subject.instance === undefined) continue
        if (!this.resolver.evaluate(subject.instance, ruleConditions, this.resolveFieldMapper(subject))) continue
        return effect === "allow"
      }

      return effect === "allow"
    }

    return false // EC-1, EC-2: default deny.
  }

  /** The subject's own `fieldMapper` (set via `createSubject`) takes precedence; `config.mapper`, keyed by the subject's resolved catalog name, is the fallback. */
  private resolveFieldMapper(subject: TSubjects): SubjectFieldMapper<unknown> | undefined {
    if (subject.fieldMapper) return subject.fieldMapper as SubjectFieldMapper<unknown>
    return this.config.mapper?.get(resolveName(this.subjectCatalog, subject.name))
  }

  private matchesAction(action: TActions, ruleAction: string, anyAction: string | typeof DISABLED): boolean {
    this.warnIfUnregisteredDynamic(action, this.actionCatalog, "Action", "createAction")
    const actionName = resolveName(this.actionCatalog, action.name)
    return actionName === ruleAction || (anyAction !== DISABLED && ruleAction === anyAction)
  }

  private matchesSubject(subject: TSubjects, ruleSubject: string, anySubject: string | typeof DISABLED): boolean {
    this.warnIfUnregisteredDynamic(subject, this.subjectCatalog, "Subject", "createSubject")
    const subjectName = resolveName(this.subjectCatalog, subject.name)
    return subjectName === ruleSubject || (anySubject !== DISABLED && ruleSubject === anySubject)
  }

  /**
   * A dynamic (no-name) Action/Subject never registered in any catalog
   * reachable from this Policy can't resolve to a real name - it falls
   * through to default-deny like any other non-match (unless a wildcard
   * rule catches it), but that's silent otherwise, so warn once per
   * distinct id rather than once per `.can()`/`.cannot()`/`.require()` call.
   */
  private warnIfUnregisteredDynamic(
    value: { name: string, __dynamic?: true },
    reverseMap: Map<string, string>,
    kind: string,
    factory: string,
  ): void {
    if (!value.__dynamic || reverseMap.has(value.name) || this.warnedDynamicIds.has(value.name)) return
    this.warnedDynamicIds.add(value.name)
    ;(this.config.logger ?? getLogger()).warn(
      `${kind} created via ${factory}() with no name was checked but never registered in any KeycardConfig catalog reachable from this Policy - it can never match a non-wildcard rule.`,
    )
  }
}
